src/types/UiaTextRangeBase.cpp

```diff
@@ -608,45 +608,55 @@ try
     });
     RETURN_HR_IF(E_FAIL, !_pData->IsUiaDataInitialized());
 
-    const std::wstring queryText{ text, SysStringLen(text) };
-    const auto bufferSize = _getOptimizedBufferSize();
-    const auto sensitivity = ignoreCase ? Search::Sensitivity::CaseInsensitive : Search::Sensitivity::CaseSensitive;
-
-    auto searchDirection = Search::Direction::Forward;
-    auto searchAnchor = _start;
-    if (searchBackward)
+    const std::wstring_view queryText{ text, SysStringLen(text) };
+    const auto results = _pData->GetTextBuffer().SearchText(queryText, ignoreCase != FALSE, _start.y, _end.y + 1);

```

Comment:
Previously, this code could use the `Search` class, because `Search` would stop searching once it found something. The new `Search` class will go through the entire buffer however.

To fix that we could either add 2 parameters to `Search` to specify a bounding range for the search or use `TextBuffer`'s function directly. I felt like the latter is a better option, because while the 2 `for` loops below are verbose and somewhat difficult to understand initially, they're also easier to verify for their correctness.

Reply:
@carlos-zamora I didn't quite understand the `UiaTextRangeBase::FindText` implementation. Could you review that file for me? 🙂 Also, how do I best test that function?

Reply:
Reviewing 😊

As for how to test it on your own, you can do that using "Accessibility Insights":
- open A11y insights
- Click on the terminal
- Under "Patterns", Click "Explore..." button next to "TextPattern"
- A "TextPattern Explorer" window should appear
- Click the "..." button next to the "Document" range
- Click "Find..."

This new window should allow you to directly interact with the API much like any screen reader could.

Reply:
It seems Accessibility Insights is kinda buggy? I can't really tell how it works, but if you open the Find dialog and ever reach a "not found" situation, it'll continue to say "not found" forever.

After testing it, I think that my implementation works correctly. If I type "eeeee" the find dialog will find all "e"s no matter where they are or whether I use forwards or backwards search. So there don't seem to be any off-by-1 errors in the code.

However, there's one big difference to the old code: My new implementation returns a "not found" result when it has reached the end of the buffer. The old code continued to circle around the buffer if you reached the end. (This didn't always work for the backward search due to the `start > _start` check. It's supposed to be `start >= _start`, otherwise it can't find results that are at the very start of the buffer at `0,0`.)
Was that intentional? Should the UIA search circle around in the buffer when it has reached either end?

Reply:
> Previously, this code could use the `Search` class, because `Search` would stop searching once it found something. The new `Search` class will go through the entire buffer however.

Won't this cause a decent performance hit? `_getOptimizedBufferSize()` used to reduce the search space by ignoring anything past the mutable bottom. Now that we're searching through the whole buffer again, won't `FindText('a', ...)` be slower. And if the user calls that function repeatedly, the user may have to wait quite a bit before being able to perform the next search? (FYI I haven't verified the performance question via a11y insights yet, so this could all be false)

> It seems Accessibility Insights is kinda buggy? I can't really tell how it works, but if you open the Find dialog and ever reach a "not found" situation, it'll continue to say "not found" forever.

You could also use [inspect.exe](https://learn.microsoft.com/en-us/windows/win32/winauto/inspect-objects). It's the predecessor to a11y insights and can be found in the Windows SDK's \\bin\\\<version\>\\\<platform\> folder. To interact with the API, do the following:
1. open inspect.exe
2. click on the terminal (the terminal should be highlighted in the UIA tree in inspect.exe)
3. Click "Action" > "Text Pattern Explorer..." (a new "Text Explorer" window should appear)
4. Click "Range" > "Find..."

> Was that intentional? Should the UIA search circle around in the buffer when it has reached either end?

Interesting! I was originally going to say that we should keep the same behavior as before, but it looks like this experience is different from MS Word's! There were two notable differences with your implementation:
1. **circling the buffer**: MS Word doesn't circle the text area whereas we do in Terminal (old and new impl). So, we probably shouldn't do that anymore.
2. **no more results**: MS Word returns something that is interpreted by inspect.exe as "ProcessFind: FindText couldn't find matching text.". But Terminal (new impl) is interpreted as "ProcessFind: MoveEndpointByRange failed. Error -2147024809". 

We should align these experiences (which is a bit annoying because your PR wasn't really intended to deal with _any_ of this). If that's easy to do, it'd be best to do that as a part of this PR or in a follow-up before the next release. If not, I think we should keep the old experience, but I know that's not really possible anymore given that the `Search` class is now completely different. Thoughts?

Reply:
> > Previously, this code could use the `Search` class, because `Search` would stop searching once it found something. The new `Search` class will go through the entire buffer however.
> 
> Won't this cause a decent performance hit? `_getOptimizedBufferSize()` used to reduce the search space by ignoring anything past the mutable bottom.

The new ICU-based searcher can search the _entire text buffer_ in 4ms.

It also avoids searching any part of the text buffer that has not been "committed" to memory. :)

Reply:
ICU also has `usearch_previous`. We could use it for UIA specifically since it doesn't need the total number of results unlike WT (which is why it uses `uregex` because it's way _way_ faster when searching for all results). Just like `uregex` it's very easy to use. I can make it fit. 🙂

And I'll try to make our UIA function behave like Word.

Reply:
Nevermind, it seems that searching through a character-iterator is a C++ API feature and Windows only ships the C API. Well it's probably fine. I feel like this is not going to be a bottleneck no matter what we do.